PARQUET-378: Add thoroughly parquet test encodings

spena opened the pull request:
A new test case TestTypeEncodings is added that test v1 and v2 encodings for all
supported column types. This test case spans many pages and row groups, and reads
each page individually from first-to-last and from last-to-first.

rdblue commented on parquet-hadoop/src/test/java/org/apache/parquet/encodings/TestTypeEncodings.java:
```diff
@@ -0,0 +1,417 @@
+
+  @Test
+  public void testTypeEncodings() throws Exception {
+    int pageSize = 64; // 64B
```
I think we should use a bigger value for this. I bumped it up to 16k and the tests work just fine. 8k or 16k is a setting more like what we would see in real data, and you can still fit several pages in each row group.

rdblue commented on parquet-hadoop/src/test/java/org/apache/parquet/encodings/TestTypeEncodings.java:
```diff
@@ -0,0 +1,417 @@
+    GroupWriteSupport.setSchema(schema, configuration);
+
+    ParquetWriter<Group> writer = new ParquetWriter<Group>(file, new GroupWriteSupport(),
+        CompressionCodecName.UNCOMPRESSED, rowGroupSize, pageSize, 0, false, false, version, configuration);
```
Could you add a comment that calls out the fact that we are avoiding dictionary encoding?

rdblue commented:
I tested this locally after removing the fix for PARQUET-246 and it caught the error. I'm +1 once the minor comments are addressed. Thanks @spena!

spena commented:
@rdblue I committed another patch addressing your feedback. I had to force the push because of problems I had.

In this patch, I added the following:
- testTypesEncodingsWithDictionary
- Increment page size from 64B to 16K
- Reduced row group size from 256K to 128K so that booleans can span more pages
- Increment record count to 2 million to span more pages
- Test every value of WriterVersion.values() in case a new one is added in the future

I run the test without the PARQUET-246 fix, and it catch the bug as well.

rdblue commented:
Thanks @spena! The tests failed, but both were due to #269. I'll commit this soon.

spena commented:
@rdblue How can I re-run the tests to see if this patch passes?

rdblue commented:
Rebase on the current master and force-push. Thanks @spena! I'll review it after.

spena commented:
@rdblue These tests added ~10 min more time to unit-testing. Would you like to lower the time spent on unit-testing, or is that ok?

rdblue commented:
Ouch. Any way we can get that down a bit? Adding 10 minutes to the build is pretty ugly. In that case I'd say we should add these as integration tests using the failsafe plugin.

spena commented:
@rdblue Now the tests run with 'mvn verify' as integration tests. Although, I checked travis-ci, and all tests run in ~17min, Last time with this encodings test were 21min. Maybe there are other tests adding more time, but this one is good for now I think.

julienledem commented:
This looks good to me.
@rdblue did you want to merge?

rdblue commented:
Merged. Thanks, Sergio!
